### ISSUE TITLE
[Question] model got unexpected prediction after saving and loading, different from the unsaved one

### ISSUE THREAD

**Issue by JaimeParker** [DLR-RM/stable-baselines3#1858]
### ❓ Question

I'm encountering a problem when **loading and using a training model, whose evaluate_policy reward mean is different from evaluating it before saving**.

First, I designed a CustomEnv based on gym.Env;

Second, I use `make_vec_env` to train the model using PPO:

```python
    vec_env = make_vec_env(CustomEnv, n_envs=10)
    model = PPO("MlpPolicy", vec_env, verbose=1)
    model.learn(total_timesteps=10000000)
    
    rew_mean, rew_std = evaluate_policy(model, model.get_env())
    print("Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)
    model.save(path)  # path is a string end with xxx.zip
    del model
```

Then I got a message:

```
---------------------------------------
| rollout/                |           |
|    ep_len_mean          | 74        |
|    ep_rew_mean          | 50        |
| time/                   |           |
|    fps                  | 3902      |
|    iterations           | 245       |
|    time_elapsed         | 2571      |
|    total_timesteps      | 10035200  |
| train/                  |           |
|    approx_kl            | 2.8975227 |
|    clip_fraction        | 0.168     |
|    clip_range           | 0.2       |
|    entropy_loss         | 1.66      |
|    explained_variance   | 0.994     |
|    learning_rate        | 0.0003    |
|    loss                 | 0.14      |
|    n_updates            | 2440      |
|    policy_gradient_loss | -0.00158  |
|    std                  | 0.106     |
|    value_loss           | 0.03  |
---------------------------------------

Mean reward per episode: 50, std of reward per episode: 0.03
```

Thirdly, I write a prediction script to test the saved model:

```python
model = PPO.load(path, env=CustomEnv())
vec_env = make_vec_env(CustomEnv, n_envs=1)
obs = vec_env.reset()

rew_mean, rew_std = evaluate_policy(model, model.get_env())
print("Using model.get_env(), Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

rew_mean, rew_std = evaluate_policy(model, vec_env)
print("Using vec_env, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

n_steps = 200
for step in range(n_steps):
    action, _states = model.predict(observation=obs)
    obs, reward, done, info = vec_env.step(action)
    if done:
    obs = vec_env.reset()
    break
```

Then I got a message:

```
Using model.get_env(), Mean reward per episode: -3.3218772999999997 , std of reward per episode 0.19078889999999996
Using vec_env, Mean reward per episode: -3.3218772999999997 , std of reward per episode 0.19078889999999996
```

Also, the action predicted doesn't satisfy the reward as trained at all.

So I tried to find out what happened when saving or loading, I re-edit the training script:

```python
    rew_mean, rew_std = evaluate_policy(model, model.get_env())
    print("Exp1, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

    model.save(path)
    rew_mean, rew_std = evaluate_policy(model, model.get_env())
    print("Exp2, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

    model2 = PPO.load(path, env=CustomEnv())
    rew_mean, rew_std = evaluate_policy(model2, model.get_env())
    print("Exp3, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

    rew_mean, rew_std = evaluate_policy(model, model2.get_env())  # error happens
    print("Exp4, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)

    rew_mean, rew_std = evaluate_policy(model2, model2.get_env())  # error happens
    print("Exp5, Mean reward per episode:", rew_mean, ", std of reward per episode", rew_std)
```

As you can see, `model` just finished training, while `model2` is loaded from where model saved. Then running the training script, I got:

```
Using cuda device
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 128      |
|    ep_rew_mean     | 0.187    |
| time/              |          |
|    fps             | 13193    |
|    iterations      | 1        |
|    time_elapsed    | 3        |
|    total_timesteps | 40960    |
---------------------------------

...

-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 74.5        |
|    ep_rew_mean          | 50.4        |
| time/                   |             |
|    fps                  | 4050        |
|    iterations           | 25          |
|    time_elapsed         | 252         |
|    total_timesteps      | 1024000     |
| train/                  |             |
|    approx_kl            | 0.008807468 |
|    clip_fraction        | 0.121       |
|    clip_range           | 0.2         |
|    entropy_loss         | -1.53       |
|    explained_variance   | 1           |
|    learning_rate        | 0.0003      |
|    loss                 | 0.0346      |
|    n_updates            | 240         |
|    policy_gradient_loss | 0.00153     |
|    std                  | 0.519       |
|    value_loss           | 0.0325      |
-----------------------------------------
Exp1, Mean reward per episode: 49.874142 , std of reward per episode 0.0
Exp2, Mean reward per episode: 49.874142 , std of reward per episode 0.0
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
Exp3, Mean reward per episode: 49.874142 , std of reward per episode 0.0
Exp4, Mean reward per episode: -1.5924555999999999 , std of reward per episode 0.21598079999999997
Exp5, Mean reward per episode: -1.520462 , std of reward per episode 0.0
```

* Exp 1,2,3 is the same as the training result
* Exp 4,5 is totally from that above

It seems that `model2.get_env()` isn't the right way, so how to fix it? Thanks.

Since the get_env() method can get a right env, I designed a test again:

```python
    vec_env = make_vec_env(CustomEnv, n_envs=1)
    obs = vec_env.reset()

    right_vec_env = make_vec_env(CustomEnv, n_envs=10)
    model_with_right_env = PPO("MlpPolicy", right_vec_env, verbose=1)
    model_with_right_env.learn(total_timesteps=int(1e6))
    right_env = model_with_right_env.get_env()

    model = PPO.load(abs_zip_path, right_env)
    print("Model loaded from ", "\033[91m{}\033[00m".format(abs_zip_path))
    rew_mean, rew_std = evaluate_policy(model, model.get_env())
    print("\033[92m{}\033[00m".format("loaded model:"), " Mean reward per episode:", rew_mean,
          ", std of reward per episode", rew_std)
```

by this way, the evaluated result is pretty close to the one just finish training.

However, the evaluated result is highly connected with `total_timesteps`, which means how goog the `model_with_right_env` have learned.

I know calling for `model_with_right_env.learn(total_timesteps=int(1e6))` is stupid when I want to do prediction, but if this line were removed, the evaluated result will be as bad as former one.

With less steps, the reward mean will also decrease. Very weird.  

**So if and only if the model saved and loaded was trained using the same total_timesteps as `model_with_right_env.learn(total_timesteps=int(1e6))`, can this model and env work correctly.** 

Does anyone know how to fix this bug? Thanks

I'm using gym 0.18.3 and gymnasium 0.28.1.

The CustomEnv is created by using gymnasium, as in [doc-Using Custom Environment](https://stable-baselines3.readthedocs.io/en/master/guide/custom_env.html)

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by maxime-louis** [comment]
Is there some stochasticity in your environment ? This may lead to variable results when calling evaluate_policy. 

If so, check carefully how to use and fix seeds in your environment.
From gymnasium 0.28:
```
def reset(
        self,
        *,
        seed: int | None = None,
        options: dict[str, Any] | None = None,
    ) -> tuple[ObsType, dict[str, Any]]:  # type: ignore
        """Resets the environment to an initial internal state, returning an initial observation and info.

        This method generates a new starting state often with some randomness to ensure that the agent explores the
        state space and learns a generalised policy about the environment. This randomness can be controlled
        with the ``seed`` parameter otherwise if the environment already has a random number generator and
        :meth:`reset` is called with ``seed=None``, the RNG is not reset.

        Therefore, :meth:`reset` should (in the typical use case) be called with a seed right after initialization and then never again.

        For Custom environments, the first line of :meth:`reset` should be ``super().reset(seed=seed)`` which implements
        the seeding correctly.

        .. versionchanged:: v0.25

            The ``return_info`` parameter was removed and now info is expected to be returned.

        Args:
            seed (optional int): The seed that is used to initialize the environment's PRNG (`np_random`) and
                the read-only attribute `np_random_seed`.
                If the environment does not already have a PRNG and ``seed=None`` (the default option) is passed,
                a seed will be chosen from some source of entropy (e.g. timestamp or /dev/urandom).
                However, if the environment already has a PRNG and ``seed=None`` is passed, the PRNG will *not* be reset
                and the env's :attr:`np_random_seed` will *not* be altered.
                If you pass an integer, the PRNG will be reset even if it already exists.
                Usually, you want to pass an integer *right after the environment has been initialized and then never again*.
                Please refer to the minimal example above to see this paradigm in action.
```
and the evaluate_policy method of stable_baselines_3 does call env.reset() before running each evaluation episode.

**Comment by JaimeParker** [comment]
@maxime-louis Thank you and it turned out to be my CustomEnv's issue, the saving and loading operation is good.

**Comment by wilhem** [comment]
> @maxime-louis Thank you and it turned out to be my CustomEnv's issue, the saving and loading operation is good.

Can you please explain, what the error was?
I stuck in the same problem and cannot find the reason
